src/coding/long.rs

```diff
@@ -7,44 +7,44 @@ pub type Long = i64;
 
 impl Decodeable<Long, io::Error> for VecDeque<u8> {
     fn decode(&mut self) -> Result<Long, io::Error> {
-        let mut result: Long = 0;
-
-        for _ in 1..=8 {
-            result += i64::from(
-                self.pop_front()
-                    .expect("Vector needs to have 8 bytes to decode a long(i64)."),
-            );
-            result <<= 8;
+        let mut temp: u64 = 0;
+
+        for i in 1..=8 {
+            let byte = self.pop_front().expect("Vector needs to have 8 bytes to decode a long(i64).") as u64;
+            temp += byte;
+            if i != 8 {
+                temp = temp << 8;
+            }
+            
         }
-
+        let msb = temp >> 63;
+        
+        let mut result: Long = temp as Long;
+        if msb == 0b1 {
+            result = -(!temp as Long);
```

Comment:
Could you explain what this does? Getting the negative from ..?

Reply:
temp is the u64 number built with the 8 bytes of the Vec. If the most significant bit is 1 (meaning the number is _originally_ negative) temp gets flipped and becomes negative.